zigzag/classes/stages/RemoveNoUseMemStage.py

```diff
@@ -0,0 +1,176 @@
+from zigzag.classes.hardware.architecture.accelerator import Accelerator
+from zigzag.classes.hardware.architecture.core import Core
+from zigzag.classes.hardware.architecture.memory_hierarchy import MemoryHierarchy
+from zigzag.utils import pickle_deepcopy
+from zigzag.classes.stages.Stage import Stage
+from typing import Generator
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+#################### Description ####################
+## This stage must be processed behind WorkloadStage.
+## This stage removes non-useful memory level found by SearchNoUseMemStage.
+################### Pseudo-code ####################
+## Initialization:
+##  target_act_mem_level, target_output_mem_level: get from mem_update_list
+##  target_const_mem_level = mem_udpate_weight
+## 1. Modify mem structure:
+## for mem in mem_levels(sort_order: from bottom to top):
+##   if ['I'] in mem.served_operand and mem.mem_level > target_act_mem_level:
+##     remove ['I'] in mem.served_operand, mem_port_alloc
+##   if ['O'] in mem.served_operand and mem.mem_level > target_output_mem_level:
+##     remove ['O'] in mem.served_operand, mem_port_alloc
+##   if ['W'] in mem.served_operand and mem.mem_level > target_const_mem_level:
+##     remove ['W'] in mem.served_operand, mem_port_alloc
+## 2. Remove no-use memory
+## for mem in mem_levels(sort_order: from top to bottom):
+##   if mem.served_operand == empty:
+##     do not add the current mem into the modified architecture
+#####################################################
+
+class RemoveNoUseMemStage(Stage):
```

Comment:
This name is not the most informative. Maybe something like "RemoveUnusedMemoryStage"?